docs/patterns/CVE-2024-27284-HIR.rpl

```diff
@@ -0,0 +1,70 @@
+pattern CVE-2024-27284
+
+util {
+    use std::iter::Iterator;
+    use cassandra_cpp_sys::cassandra::case_iterator_next;
+    /*
+        All function below has property: 
+        - get an item from the iterator `*const CassIterator`
+        - 'Calling cass_iterator_next() will invalidate the previous'
+    */
+    use cassandra_cpp_sys::cassandra::{
+        cass_iterator_get_row,
+        cass_iterator_get_column,
+        cass_iterator_get_value,
+        cass_iterator_get_map_key,
+        cass_iterator_get_map_value,
+        cass_iterator_get_user_type_field_name,
+        cass_iterator_get_user_type_field_value,
+        cass_iterator_get_keyspace_meta,
+        cass_iterator_get_table_meta,
+        cass_iterator_get_materialized_view_meta,
+        cass_iterator_get_user_type,
+        cass_iterator_get_function_meta,
+        cass_iterator_get_aggregate_meta,
+        cass_iterator_get_column_meta,
+        cass_iterator_get_index_meta,
+        cass_iterator_get_meta_field_name,
+        cass_iterator_get_meta_field_value,
+    }
+    /*
+        pub enum cass_bool_t {
+            cass_false = 0,
+            cass_true = 1,
+        }
+    */
+    use cassandra_cpp_sys::cassandra::{cass_false, cass_true};
+    use std::option::Option;
+
+    p_non_idiomatic_use_of_iterator = 
+        impl Iterator for $T {
+            fn next(&mut self) -> Option<_> {
+                unsafe {
+                    match cass_iterator_next($iter) {
+                        cass_false => None,
+                        cass_true => {
+                            let $item = #or!{
```

Comment:
Is this expression reasonable? Is there a better way?
@frank-king @TheVeryDarkness

Reply:
Considier the #! 'macro' as a String Substitution